File: 0695-max-area-of-island/0695-max-area-of-island.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int area = 0;
        int res = 0;
        function<void(int, int)> dfs = [&dfs, &grid, m, n, &area](int i, int j) {
            if (i >= 0 && i < m && j >=0 && j < n && grid[i][j]) {
                grid[i][j] = 0;
                area++;
                dfs(i+1,j);
                dfs(i-1,j);
                dfs(i,j+1);
                dfs(i,j-1);
            }
        };

        for(int i=0;i<m;i++) {
            for(int j=0;j<n;j++) {
                if(grid[i][j]) {
                    area = 0;
                    dfs(i,j);
                    res = res > area? res:area;
                }
            }
        }
        return res;
        
    }
};

File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        self.area = []
        self.m = len(grid)
        self.n = len(grid[0])
        def dfs(i:int,j:int):
            if 0 <= i < self.m and 0 <= j < self.n and grid[i][j]:
                self.area[-1]+=1
                grid[i][j]=0
                dfs(i+1,j)
                dfs(i-1,j)
                dfs(i,j+1)
                dfs(i,j-1)
        
        for i in range(self.m):
            for j in range(self.n):
                if grid[i][j]==1:
                    self.area.append(0)
                    dfs(i,j)
        
        if self.area:
            return max(self.area)
        else:
            return 0

        